northstar/src/runtime/ipc/channel.rs

```diff
@@ -1,93 +1,218 @@
 use std::{
-    io::{ErrorKind, Read, Write},
+    io::{Read, Write},
     os::unix::prelude::{AsRawFd, RawFd},
 };
 
 use bincode::Options;
 use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
 use serde::{de::DeserializeOwned, Serialize};
-use tokio::io::AsyncReadExt;
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
 
-use super::pipe::{pipe, AsyncPipeRead, PipeRead, PipeWrite};
+use super::pipe::{pipe, AsyncPipeRead, AsyncPipeWrite, PipeRead, PipeWrite};
 
-/// Wrap a pipe to transfer bincoded structs that implement `Serialize` and `Deserialize`
-pub struct Channel {
-    tx: PipeWrite,
-    rx: std::io::BufReader<PipeRead>,
+/// [`PipeWrite`] wrapper for sending serializable data
+pub struct Sender<S> {
+    pipe: PipeWrite,
+    _marker: std::marker::PhantomData<S>,
 }
 
-impl Channel {
-    /// Create a new pipe channel
-    pub fn new() -> Channel {
-        let (rx, tx) = pipe().expect("Failed to create pipe");
-        Channel {
-            tx,
-            rx: std::io::BufReader::new(rx),
+impl<S> Sender<S>
+where
+    S: Serialize + DeserializeOwned,
+{
+    /// Send bincode serialized
+    pub fn send(&mut self, s: S) -> std::io::Result<()> {
+        let buffer = bincode::DefaultOptions::new()
+            .with_fixint_encoding()
+            .serialize(&s)
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
+        self.pipe.write_u32::<BigEndian>(buffer.len() as u32)?;
+        self.pipe.write_all(&buffer)
+    }
+}
+
+impl<S> AsRawFd for Sender<S> {
+    fn as_raw_fd(&self) -> RawFd {
+        self.pipe.as_raw_fd()
+    }
+}
+
+impl<S> From<PipeWrite> for Sender<S>
+where
+    S: Serialize + DeserializeOwned,
+{
+    fn from(pipe: PipeWrite) -> Self {
+        Sender {
+            pipe,
+            _marker: std::marker::PhantomData,
         }
     }
+}
+
+/// [`PipeRead`] wrapper for receiving serializable data
+pub struct Receiver<S> {
+    pipe: std::io::BufReader<PipeRead>,
+    _marker: std::marker::PhantomData<S>,
+}
 
-    /// Raw fds for the rx and tx pipe
-    pub fn as_raw_fd(&self) -> (RawFd, RawFd) {
-        (self.tx.as_raw_fd(), self.rx.get_ref().as_raw_fd())
+impl<S> Receiver<S>
+where
+    S: Serialize + DeserializeOwned,
```

Comment:
Why is `Serialize` needed here?

Reply:
Is not _needed_ for only the receiver implementation but in general a pair of `Sender` `Receiver` will share the same type and therefore it has to implement both `Serialize` and `Deserialize`. That's the reason i set these bounds everywhere.

Reply:
Makes sense.